### PR TITLE
Solution for Chapter 4

### PR DESCRIPTION
**Pull request by khvmaths**
### Solutions for Chapter 4

Thank you once again! You guys are awesome. 😎  Indeed, I admit that I still a bit confuse about everything starting Chapter 3, but I believe practice make perfects!

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter4.hs [hunk 1]
```diff
@@ -307,6 +308,11 @@ data List a
     = Empty
     | Cons a (List a)
 
+instance Functor List where
+  fmap :: (a -> b) -> List a -> List b
+  fmap _ Empty = Empty
+  fmap f (Cons x xs) = Cons (f x) (fmap xs)
```
Comment (vrom911) [inline src/Chapter4.hs:314]: Great recursive implementation 👍🏼

File: src/Chapter4.hs [hunk 2]
```diff
@@ -472,10 +478,11 @@ Implement the Applicative instance for our 'Secret' data type from before.
 -}
 instance Applicative (Secret e) where
     pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+    pure = Reward
 
     (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+    Trap e <*> _ = Trap e
+    Reward f <*> x = fmap f x
```
Comment (vrom911) [inline src/Chapter4.hs:485]: Nice work on reusing `Functor` instance for `Secret`! :world_map:

File: src/Chapter4.hs [hunk 3]
```diff
@@ -488,6 +495,15 @@ Implement the 'Applicative' instance for our 'List' type.
   may also need to implement a few useful helper functions for our List
   type.
 -}
+instance Applicative List where
+  pure :: a -> List a
+  pure x = Cons x Empty
+
+  (<*>) :: List (a -> b) -> List a -> List b
+  Empty <*> _ = Empty
+  _ <*> Empty = Empty
```
Comment (vrom911) [inline src/Chapter4.hs:504]: Nice catch, this case makes this implementation much more efficient! 🚤

File: src/Chapter4.hs [hunk 4]
```diff
@@ -611,6 +628,13 @@ Implement the 'Monad' instance for our lists.
   maybe a few) to flatten lists of lists to a single list.
 -}
 
+instance Monad List where
+  (>>=) :: List a -> (a -> List b) -> List b
+  l >>= f = concatList(fmap f l)
```
Comment (vrom911) [inline src/Chapter4.hs:633]: Great!
Nothe, that in Haskell, function are separated from arguments with spaces 🙂 :
```suggestion
  l >>= f = concatList (fmap f l)
```

File: src/Chapter4.hs [hunk 5]
```diff
@@ -629,7 +653,7 @@ Can you implement a monad version of AND, polymorphic over any monad?
 🕯 HINT: Use "(>>=)", "pure" and anonymous function
 -}
 andM :: (Monad m) => m Bool -> m Bool -> m Bool
-andM = error "andM: Not implemented!"
+andM aa bb = aa >>= \a -> if a then bb else pure False
```
Comment (vrom911) [inline src/Chapter4.hs:656]: Yes! This is the correct lazy monadic implementation 👍🏼

**Review by vrom911 — APPROVED** [review]
Awesome implementation of the Chapter! Nice work on the Tree task as well!

Thanks a lot for taking this course ❤️ It was a pleasure to see your creative solutions 👍🏼 
If you have any questions or would like to discuss anything we are always here! 🤗 
Also, we appreciate [any feedback](https://docs.google.com/forms/d/e/1FAIpQLScBVhLxq5CgGnAfIGUE-fCoOUqeGkDY2HXzbT7KV2jjLOsmjQ/viewform)! 🙏🏼

**Comment by chshersh** [comment]
I'm going to close this PR because we can't merge the solutions to our repo :slightly_smiling_face: 
Feel free to open PRs with solutions to other chapters!

**Pull request closed without merging**